docs/adr/0009-immutable-keys.md

```diff
@@ -0,0 +1,46 @@
+# 9. Immutable Application and Handler Keys
+
+Date: 2019-06-28
+
+## Status
+
+Accepted
+
+## Context
+
+Engine implementations require a mechanism for associating ancillary data with
+Dogma applications and handlers.
+
+For example, such data might include application state in the form of aggregate
+and process roots, or historical events in an event sourcing system.
+
+Currently, engine implementations rely on application and handler names as a key
+for associated data. This is especially problematic for handlers as the name
+initially chosen for a handler may become misleading over time as the handler's
+implementation changes.
+
+## Decision
+
+We've decided to add an additional identifier to applications and handlers
+called the "key".
+
+The key's express purpose is for identifying associated data, and as such as
+more stringent requirements on its immutability.
```

Comment:
Something seems a bit off here.
> and as such as

Reply:
It was supposed to be "and as such **has**", but "as such" is not appropriate here anyway. Reworded.